Replaced LINQ expressions in the Immediate Renderer with ArrayPools t…

AndrejBunjac opened the pull request:
Immediate renderer does a lot of per frame allocations - especially in the HitTest.

This fix is aimed at replacing the main culprits for this behavior - LINQ expressions.
Both Render and HitTest methods of ImmediateRenderer use .OrderBy and .Select - 
these were replaced by taking ArrayPool resources to stabilize per frame memory usage as well as creating StableSort replacements (that would work the same way as OrderBy).

VisualExtensions - added 
* StableSort(this IVisual[] values, int length, IComparer<IVisual> comparer)
* StableHitTestSort(this IVisual[] values, int length)
* StabilizingComparer : IComparer<KeyValuePair<int, IVisual>> nested class

Added new comparer - ZOrderHitTestComparer for KeyValuePair<int, IVisual> to replace the usage of ZOrderElement arrays that allocate.

ImmediateRenderer - edited HitTest and Render methods to use all of the above with ArrayPools.

TODO: Did not remove the OrderByZIndex because removing it was failing a contract. Needed feedback for that.
HitTest still allocates due to the creation of IEnumerable and might need a more robust solution that would require a system redisgn.

Measured the outcome with dotMemory - per frame allocations of Render are completely gone from HitTest are nearly halved.

AndrejBunjac committed:
Replaced LINQ expressions in the Immediate Renderer with ArrayPools to reduce per frame allocations. Added a stable sort and a hit test comparison with a new comparer to visual extensions to compensate for lack of LINQ.

kekekeks commented on src/Avalonia.Visuals/Rendering/ImmediateRenderer.cs:
```diff
@@ -250,13 +251,35 @@ private static Rect GetTransformedBounds(IVisual visual)
+                    if (childrenCount > 1)
                     {
-                        foreach (var result in HitTest(child, p, filter))
+                        var childrenArray = ArrayPool<IVisual>.Shared.Rent(childrenCount);
```
1) ArrayPool can return an array larger than requested, so some elements will be null
2) try/copy/finally/return pattern is too bloated

Consider using `PooledList<T>` instead and probably adding a constructor that takes an existing IReadOnlyList<T> to copy contents from.

AndrejBunjac replied:
I have a question about this. I already account for ArrayPool adding extra elements by iterating with for instead of foreach. However, looking at PooledList<T>, I see that creating it every time might also extensively allocate on it's own cos the object itself consists of multiple ints and pointers that will have to be allocated every time this is created. It will probably be beneficial when multiple children exist but I think that a more common use case is that IVisual has a few children in which case I think using ArrayPool directly actually significantly helps with performance.

AndrejBunjac replied:
I have a question about this. I already account for ArrayPool adding extra elements by iterating with for instead of foreach. However, looking at PooledList, I see that creating it every time might also extensively allocate on it's own cos the object itself consists of multiple ints and pointers that will have to be allocated every time this is created. It will probably be beneficial when multiple children exist but I think that a more common use case is that IVisual has a few children in which case I think using ArrayPool directly actually significantly helps with performance.

AndrejBunjac replied:
So is the list itself possibly getting cached somewhere?

MarchingCube replied:
I think using `PooledList` here can cause extra allocations that will be hard to avoid. My primary worry is that with deep control hierarchies (`TreeView` in ControlCatalog for example) default pooling stops working completely and adds extra overhead.

This is caused by a limit of 8 pooled arrays per size (and per core) so when rendering and reaching lower depths pooled arrays will be allocated and discarded after use. We could investigate if we can rewrite current recursion into iteration so one shared render/hit test pool can be kept.

AndrejBunjac replied:
I addressed these problems in the following manner:

I created two Renderer-wise PooledLists<(int, IVisual)> - one for hit testing, the other for rendering.

The reason why they hold int is to be able to not allocate for sorting and just use a pre-allocated sorter from the ImmediateRenderer (new).

The int is there to keep the original position in the parent's visualChildren and is used for stable sorting.

The pooled lists are used as sort of stacks where ranges are added and removed once the recursion branch is finished with them.

I also reworked the hit test to have separate implementations for the HitTestFIrst that was allocating way too much when it just needs to return the first element hit. For the HitTest that returns IEnumerable I actually reverted to the old implementation.

The measurement after this show 0 per frame allocations from the ImmediateRenderer.

kekekeks replied:
>However, looking at PooledList, I see that creating it every time might also extensively allocate on it's own cos the object itself consists of multiple ints and pointers that will have to be allocated every time this is created.

You can use ObjectPool to pool PooledList instances.

AndrejBunjac replied:
I redesigned how I approach the whole idea so I don't think that such pooling is needed now. I use only two PooledList objects per lifetime of the ImmediateRenderer.

AndrejBunjac committed:
Introduced pooled list to immediate renderer

AndrejBunjac committed:
Removed unnecessary iteration that allocates from ClearTransformedBounds.

AndrejBunjac commented:
Copying the same comment from the conversation above here:

I addressed these problems in the following manner:

I created two Renderer-wise PooledLists<(int, IVisual)> - one for hit testing, the other for rendering.

The reason why they hold int is to be able to not allocate for sorting and just use a pre-allocated sorter from the ImmediateRenderer (new).

The int is there to keep the original position in the parent's visualChildren and is used for stable sorting.

The pooled lists are used as sort of stacks where ranges are added and removed once the recursion branch is finished with them.

I also reworked the hit test to have separate implementations for the HitTestFIrst that was allocating way too much when it just needs to return the first element hit. For the HitTest that returns IEnumerable I actually reverted to the old implementation.

The measurement after this show 0 per frame allocations from the ImmediateRenderer.

AndrejBunjac committed:
Reverted the change to ClearTransformedBounds and some minor code style fixes

AndrejBunjac commented:
Minor fixes to code style added.

grokys commented:
Sorry this PR has gone unattended for so long @AndrejBunjac.

@kekekeks @maxkatz6 is it worth merging this PR at this point given that the immediate renderer is about to be removed?

maxkatz6 commented:
Closing as Immediate Renderer and Deferred Renderer are planned to be removed in 11.0.
